File: 프로그래머스/1/250121. ［PCCE 기출문제］ 10번 ／ 데이터 분석/［PCCE 기출문제］ 10번 ／ 데이터 분석.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int changeNum(string str) {
    if (str == "code") return 0;
    else if (str == "date") return 1;
    else if (str == "maximum") return 2;
    else return 3; // remain의 경우
}

struct Compare {
    int index;  
    Compare(int idx) : index(idx) {}  

    bool operator()(const vector<int>& a, const vector<int>& b) const {
        return a[index] < b[index]; 
    }
};

vector<vector<int>> solution(vector<vector<int>> data, string ext, int val_ext, string sort_by) {
    vector<vector<int>> answer;
    int ext_num = changeNum(ext);
    int sort_num = changeNum(sort_by);
 
    for (const auto& row : data) {
        if (row[ext_num] < val_ext) {
            answer.push_back(row);
        }
    }
    sort(answer.begin(), answer.end(), Compare(sort_num));
    
    return answer;
}

File: 프로그래머스/1/250121. ［PCCE 기출문제］ 10번 ／ 데이터 분석/［PCCE 기출문제］ 10번 ／ 데이터 분석.java
import java.util.*;

class Solution {
    public int[][] solution(int[][] data, String ext, int val_ext, String sort_by) {
        
        ArrayList<int[]> filtered = new ArrayList<>();
        
        int n = data.length;
        int extidx = changeNum(ext);
        int sort = changeNum(sort_by);
        
        for(int[] d : data){
            if(d[extidx] < val_ext){
                filtered.add(d);
            }
        }
        
       filtered.sort(Comparator.comparingInt(o ->o[sort]));
        
        int[][] answer = new int[filtered.size()][];
        for(int i =0; i<filtered.size(); ++i){
            answer[i] = filtered.get(i);
        }
        

        
        return answer;
    }
    
    private int changeNum(String ext){
        if(ext.equals("code")){
            return 0;
        }
        else if(ext.equals("date")){
            return 1;
        }
        else if(ext.equals("maximum")){
            return 2;
        }
        else{ // remain
            return 3;
        }
    }
}